### PR TITLE
Add eval reproducibility artifacts (RFC #880)

### LINKED ISSUE #880: [RFC] Add reproducibility artifacts to eval runs and checkpoints
## Motivation

Evaluation results are currently not reproducible from their output alone. The eval script (`scripts/evaluate/evaluate.py`) produces only benchmark data (`acceptance.csv`, `perf_results.csv`, raw guidellm JSONs) — no metadata about the environment, command, or code state that produced them.

This caused unnecessary re-work in #757 where we couldn't determine which vLLM patches had been applied to original eval runs. Training already records provenance via `save_train_command()` — evaluation has no equivalent.

## Artifact Summary

| Artifact | Script | Default Location | Format |
|---|---|---|---|
| `train_command.txt` | `train/utils.py` | `<checkpoint_dir>/` | `shlex.join(sys.argv)` |
| `eval_command.txt` | `evaluate.py` | `<output_dir>/` | Header (timestamp, git SHA, package versions) + `shlex.join(sys.argv)` |
| `vllm_command.txt` | `launch_vllm.py` | `vllm_<model>_<ts>/` | Header (timestamp, python, git SHA, vllm version) + `shlex.join(cmd)` |
| `vllm.patch` | `launch_vllm.py` | `vllm_<model>_<ts>/` | `# repo: <path> (<sha>)` + `git diff HEAD` |
| `checkpoint_sha256.txt` | `launch_vllm.py` | `vllm_<model>_<ts>/` | `<sha256>  <filename>` per `.safetensors` file |
| `speculators.patch` | `train/utils.py` | `<checkpoint_dir>/` | `git diff HEAD` |

`launch_vllm.py` provenance location is overridable with `--provenance-dir` to co-locate with eval results.

### 1. `eval_command.txt` — eval metadata file

Mirror the existing `save_train_command()` pattern. Write to the eval output directory:
- Timestamp (UTC ISO 8601)
- Speculators git SHA
- Package versions (speculators, vllm, transformers, torch, compressed-tensors)
- Full `sys.argv` eval command

**Integration point:** `run_benchmark()` in `evaluate.py`, right after `artifacts_dir.mkdir()`.

### 2. `vllm_command.txt` + `vllm.patch` + `checkpoint_sha256.txt` — vLLM provenance

Written by `launch_vllm.py` (not the eval script) to keep concerns separated — the eval script doesn't know how vLLM was launched.

- `vllm_command.txt`: full constructed command + metadata header (timestamp, python path, git SHA, vllm version)
- `vllm.patch`: `git diff HEAD` from the vLLM source tree (auto-discovered by walking up from the installed package, falling back to `~/vllm` and `/workspace/vllm`). Header-only when the checkout is clean. Records version string when installed from a wheel.
- `checkpoint_sha256.txt`: SHA256 of all `.safetensors` files for local checkpoints, or a comment noting the model is a remote reference.

`launch_vllm.py` always saves provenance — defaults to `vllm_<model>_<YYYYMMDD_HHMMSS>/`, overridable with `--provenance-dir`.

### 3. `speculators.patch` — speculators diff artifact

Add `git diff HEAD` from the speculators repo root to `<checkpoint_dir>/` alongside the existing `train_command.txt`. This captures uncommitted changes at training time so the diff travels with the checkpoint (useful for HF uploads).

## Implementation Notes

- All artifacts are best-effort (wrapped in try/except) — failure to capture provenance never blocks training or evaluation.
- Atomic writes via tempfile + rename, matching the existing `save_train_command()` pattern.
- For wheel installs (no git checkout), the version string is recorded and no diff is available.

### PR DESCRIPTION
**Pull request by orestis-z**
## Summary
- **`evaluate.py`**: writes `eval_command.txt` (timestamp, git SHA, package versions, full command) into the eval output directory at the start of each run.
- **`launch_vllm.py`**: always writes three provenance artifacts to `vllm_<model>_<timestamp>/` (overridable with `--provenance-dir`):
  - `vllm_command.txt` — full launch command + metadata (timestamp, python path, git SHA w/ dirty marker, vllm version)
  - `vllm.patch` — git diff of the vllm checkout (header-only for clean checkouts, version note for wheel installs)
  - `checkpoint_sha256.txt` — SHA256 of all `.safetensors` files for local checkpoints, or model name for HF hub models

Closes #880

## Test plan
- [ ] `make quality` passes
- [ ] Run `launch_vllm.py <model>` and verify `vllm_<model>_<ts>/` is created with `vllm_command.txt`, `vllm.patch`, and `checkpoint_sha256.txt`
- [ ] Run `launch_vllm.py <model> --provenance-dir /tmp/custom` and verify artifacts land in `/tmp/custom/`
- [ ] Run eval and verify `eval_command.txt` appears in output dir
- [ ] Test with wheel-installed vllm (no git repo) — `vllm.patch` should contain version note
- [ ] Test with HF hub model — `checkpoint_sha256.txt` should contain model name

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by shotsan — COMMENTED** [review]
This looks excellent and handles all the edge cases (like missing safetensors files or remote HF models). I especially appreciate the comprehensive unit tests for `launch_vllm.py`. Looks great to me! Sorry for duplicating this in #910, I missed this PR earlier.

**Comment by orestis-z** [comment]
> This looks excellent and handles all the edge cases (like missing safetensors files or remote HF models). I especially appreciate the comprehensive unit tests for `launch_vllm.py`. Looks great to me! Sorry for duplicating this in #910, I missed this PR earlier.

No worries! Thanks for your feedback!

File: scripts/evaluate/evaluate.py [hunk 1]
```diff
@@ -90,6 +95,66 @@ def _sanitize_dir_name(name: str) -> str:
...
+        tmp_path = Path(tmp)
+        try:
+            with os.fdopen(fd, "w") as f:
+                f.write(content)
+            tmp_path.replace(dest)
+        finally:
+            if tmp_path.exists():
+                tmp_path.unlink()
+    except Exception:
+        logger.warning("Failed to save eval_command.txt", exc_info=True)
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply scripts/evaluate/evaluate.py:177]: That would affect only eval code itself, this is a narrow edge-case not worth addressing

**Review by shanjiaz — COMMENTED** [review]
@orestis-z I'm a little confused by the PR. `launch_vllm.py` is specifically written for hidden states extraction. I might be misunderstanding something here, is this intended to use `launch_vllm.py` to spin up an inference instance for evaluation? I agree it would be great to add reproducibility artifacts for eval tho.

**Comment by orestis-z** [comment]
Superseded by #957 (eval provenance) and #958 (vLLM launch provenance). The original PR has been split into smaller, focused PRs.
